Gutenberg: Add Blog Stats Block

Aurorum opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes #17588

## Proposed changes:
Converts the Blog Stats legacy widget into a block so that it can be used with FSE. This is broadly just a 1:1 convert of a pretty simple (albeit popular) widget, but I've added a feature whereby users can display views for a single item rather than their entire site.

<img width="1583" alt="Screenshot 2023-12-09 at 19 09 22" src="https://github.com/Automattic/jetpack/assets/43215253/7b89b5be-9abd-46c1-b71f-4667f46fbd82">

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
See the original issue and https://github.com/Automattic/wp-calypso/issues/63584#issuecomment-1845865418

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No - it uses Stats data, but in the same way as the Legacy Widget does. 

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

You'll need to enable Beta blocks for this. When you insert the Blog Stats block, you should be able to insert text that follows the view count, similar to the widget. I thought that it was a nice touch to display the actual view count where possible, but "12,345" should appear as a placeholder when there is no Post ID (eg. widgets/FSE templates). The correct view count should appear when on the front end, so you might want to confirm this happens. 

<img width="1527" alt="Screenshot 2023-12-09 at 22 34 38" src="https://github.com/Automattic/jetpack/assets/43215253/5efc52d5-710b-47a7-90a1-448a2a64e1cf">

Also, you'll want to verify that you're encouraged to active the Stats module if it's disabled, and that setting successfully works. I felt it was best not to share the code with the Top Posts block here because it's not that much, and I've also stressed that views are only counted when the module is active, which feels more relevant to this block. 

For context, this is the legacy widget it replaces - it should be a fairly 1:1 conversion, besides the extra feature to display stats for an individual post:
<img width="686" alt="Screenshot 2023-12-09 at 19 13 51" src="https://github.com/Automattic/jetpack/assets/43215253/8a787f6a-599f-475d-a077-22db119154ff">

@kangzj super sorry to ping you for another stats block so soon - if you do fancy taking a look at this one (and totally understand if not given you've just reviewed the Top Posts and Pages block!), this one should be 10x simpler :)

Aurorum committed:
Initial commit

Aurorum committed:
Add block

Aurorum committed:
Improving handling for no post ID

Aurorum committed:
Improve i18n

Aurorum commented on projects/plugins/jetpack/modules/widgets/blog-stats.php:
```diff
@@ -99,7 +111,7 @@ public function form( $instance ) {
 			<input class="widefat" id="<?php echo esc_attr( $this->get_field_id( 'hits' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name( 'hits' ) ); ?>" type="text" value="<?php echo esc_attr( $instance['hits'] ); ?>" />
 		</p>
-		<p><?php esc_html_e( 'Hit counter is delayed by up to 60 seconds.', 'jetpack' ); ?></p>
+		<p><?php esc_html_e( 'Hit counter is delayed by up to 5 minutes.', 'jetpack' ); ?></p>
```
The caching time has increased since the widget was released (I think it happened when moving functions to the Stats package)

https://github.com/Automattic/jetpack/blob/57221b8fc8d1dcd09b00d956b873c4464205eb8f/projects/packages/stats/src/class-wpcom-stats.php#L31-L36

kangzj commented:
Wow another productive week uh @Aurorum 🤩  Nice one! I'll take a spin this week. Thank you so much for your contribution ❤️

Aurorum commented:
Thanks so much! No rush at all. :)

jeherve commented on projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-blog-stats.php:
```diff
@@ -0,0 +1,81 @@
+	public function get_blog_stats( $request ) {
+		$post_id   = $request->get_param( 'post_id' );
+		$post_data = convert_stats_array_to_object(
+			( new WPCOM_Stats() )->get_post_views( $post_id, array( 'fields' => 'views' ) )
```
The one problem we'll face with this implementation is that this will blow up folks' options table when they use the block in a page template. 

Every time we fetch post views for a post, we cache the data in a transient. If we start regularly fetching views for all posts on a site, this will make for a lot of transients. I'm not sure we'd want that to happen.

cc @Automattic/red for their opinion on this.

Aurorum replied:
Ah, that's a very good point - I didn't think of that! What do you think about batching the post views together in a single transient? Something like this:

```php
/**
 * Fetches stats data from WPCOM or local Cache. Caches locally for 5 minutes.
 *
 * Unlike the above function, this caches post data together. It's intended to
 * prevent wp_options from blowing up when retrieving views for a large number
 * of posts at the same time. However, the final response is the same as above.
 *
 * @param array $args Optional query parameters.
 * @param int   $post_id Post ID to acquire stats for.
 *
 * @return array|WP_Error
 */
protected function fetch_post_stats( $args = array(), $post_id ) {
	$endpoint       = $this->build_endpoint();
	$transient_name = self::STATS_CACHE_TRANSIENT_PREFIX . 'batched_posts';
	$stats_cache    = get_transient( $transient_name );

	if ( ! $stats_cache ) {
		$stats_cache = array();
		$expiration  = self::STATS_CACHE_EXPIRATION_IN_MINUTES * MINUTE_IN_SECONDS;
	} else {
		$expiration = get_option( '_transient_timeout_' . $transient_name ) - time();
		$data       = $stats_cache[ $post_id ];

		if ( isset( $data ) ) {
			if ( is_wp_error( $data ) ) {
				return $data;
			}

			return array_merge(
				array( 'cached_at' => time() - $expiration ),
				(array) json_decode( $data, true )
			);
		}
	}

	$wpcom_stats = $this->fetch_remote_stats( $endpoint, $args );

	$stats_cache[ $post_id ] = is_wp_error( $wpcom_stats ) ? $wpcom_stats : wp_json_encode( $wpcom_stats );

	set_transient( $transient_name, $stats_cache, $expiration );

	return $wpcom_stats;
}
```

The problem that I can envisage is the transient holding a lot of information which slows things down - but since it resets every five minutes, I don't think that should be a big issue. The other downside though is that not all posts will be cached for five minutes (some will be cached for less), but I still think that's better than no caching at all. 

If you have other ideas, please let me know!

jeherve replied:
I'm afraid this would be problematic on sites with a lot of posts. 

I'll let folks from @Automattic/red chime in on this. Unfortunately this is a tricky problem.

kangzj replied:
Thanks for working on this! It looks amazing @Aurorum 

Thanks for your feedback @jeherve 

This is a tricky indeed. Additionally, the block could possibly have impact on user's page performance, as the data needs to be refreshed every 5 mins. So we really want to be really cautious about it.

Could we cache in post meta table? Would that potentially blow it up as well?

Another thought is we could go totally async calls to API without any cache - frontend loads the views from an API which doesn't cache data on the site.

looping in more crew friends for advice @Automattic/jetpack-crew 😄

Aurorum replied:
Thanks so much @kangzj! Those both sound like viable alternatives to me. 

Let me know if anyone has a preference, and I’ll try to include it here.

jeherve replied:
> we could go totally async calls to API without any cache - frontend loads the views from an API which doesn't cache data on the site.

👍 This may be the lightest solution, it wouldn't clog databases and shouldn't have as big an impact on the frontend.

Aurorum replied:
Thanks! I've added that now. :)

anomiex replied:
> 👍 This may be the lightest solution, it wouldn't clog databases and shouldn't have as big an impact on the frontend.

As long as it doesn't result in every frontend page view hitting the API or database. If that happens, it'll probably wind up tripping some WordPress.com monitor and we'll have to add some sort of caching back in.

jeherve replied:
Hm, good point.

Aurorum replied:
If it’s more trouble than it’s worth, I can always just remove the feature - seems a popular request (and a bit defeatist!!) though.

Is there a downside to caching in the post meta table?

Aurorum replied:
I've gone with @kangzj's idea of caching in the post meta since it actually looks like that's what the Related Posts feature already does, and it seems to resolve the problem here. 

Let me know if I should change it!

Aurorum committed:
Merge remote-tracking branch 'upstream/trunk' into add/stats-block

Aurorum committed:
Override cache

Aurorum committed:
Run tools/fixup-project-versions.sh

Aurorum committed:
Merge branch 'trunk' into add/stats-block

Aurorum committed:
Cache stats again

Aurorum committed:
Cache in post meta

Aurorum commented:
Trying to fix the tests, but I seem to get this error whenever I try to update the branch (not just this branch, but any Jetpack one).

```
Checking for filename collisions. Just a sec...
No collisions detected.
Checking if changelog files are needed. Just a sec...
Project packages/backup-helper-script-manager is being changed, but no change file in projects/packages/backup-helper-script-manager/changelog/ is touched!
Use `jetpack changelogger add <slug>` to add a change file for each project.
Guidelines: https://github.com/Automattic/jetpack/blob/trunk/docs/writing-a-good-changelog-entry.md
Checking if changelog files are needed. Just a sec...
? Found 46 project(s) that need a changelog. Create and add the same changelog to all of them?
```

Any ideas what I'm doing wrong?! I'm assuming that I'm not supposed to actually add a changelog for each project.

kraftbj commented:
At first blush, I thought it might be a situation where if you update a package version that is used in a lot of different packages, there will often need a changelog in all of the impacted projects to acknowledge the package dependencies were updated.

But, I don't think that's what's happening here. Or at least not as straightforward. It looks like your branch is a little behind trunk as well, I opened https://github.com/Aurorum/jetpack/pull/2 for you to pull in the latest trunk changes. I'm curious if you accept that PR into your branch, if the issue will continue.

kraftbj committed:
Catch up your branch (#2)

* Social: Fix auto-conversion cleanup logic (#34666)

* Fix logic to be mindful of absent option

* changelog

* Fixup versions

* Launchpad: Improve the WC task visibility check (#34648)

* Improve the WC task visibility check

* Bump version

* Fix composer

* changelog

* Bump version

---------

Co-authored-by: Valter Lorran <valter.lorran@automattic.com>

* Improve sharing buttons events performance (#34652)

* Backport mu-wpcom-plugin 2.0.7, jetpack 13.0-a.5 Changes (#34676)

* Changelog edits.

* Init new cycle

* Carousel: fix unresponsive navigation (#34678)

* Check for both aksimet and antispam for backwards compatibility (#34674)

* Jetpack Google Fonts: Fix the fonts provided by theme are still added (#34608)

* Google Fonts: Ignore the fonts that are provied by the active theme

* changelog

* Update syntax

* Subscribers: fix the reach count above 1000 (#34689)

Allows strings in translation strings. Ensures reachCount is consistently a Number, and convert it to string for localization.

* Subscriptions: localize number format in access panel (#34691)

* Added preemptive check full sync posts processing (#34661)

* Added preemptive check to break the loop of adding metadata if we are already over the MAX SIZE

* Added changelog

* Added break 2 since there is no point in doing the comparison in the outer loop which will break too

* Version bump

* Like block: remove like module dependency (#34664)

* - Remove dependency on likes module
- Render master iframe (only once)

* Changelog

* Don't render anything when post_id is missing

* Better check

* Change master to main

* Enqueue styles & register them on first usage

* Remove require_once

* Update iframe for WPCOM

* Require the module

* Docs: update quick start guide (#34656)

* Clarify quick start guide is geared for a12s

* Cleanup cloning and install script steps

* Move env checker script heading below install steps

* Adjust heading levels

* Move jtube a11n-only note to top of its section

* Add ngrok/jtube info to full dev-env guide

* CRM: Ensure segments can be deleted (#34690)

* Update JS unit testing packages (major) (#34694)

* Update JS unit testing packages
* Update indirect @adobe/css-tools dep

---------

Co-authored-by: Renovate Bot <bot@renovateapp.com>
Co-authored-by: Brad Jorsch <brad.jorsch@automattic.com>

* Update storybook monorepo to v7.6.5 (#34696)

Co-authored-by: Renovate Bot <bot@renovateapp.com>

* Show an error when one is encountered while rendering a payment button (#34380)

* Render an error message when the button cannot be rendered

* changelog

* CRM: Resolves 3399 - make KB links consistent in settings (#34695)

* Adjust wording to clarify all fields can show in Woo My Account

* Add KB link

* Use consistent UI to link to KB from settings

* Simplify wording

* Remove stray end tag

* Remove errant colspan attribute

* Add changelog

* Only contact fields

* Feature: AI Assistant UI fixes (#34383)

* use spinner instead of ai icon

* adjust main action button styles

* only show main action buttons when there's content. Switch between icon and label for mobile. Do not show clear button is state is requesting

* do not show clearbutton

* changelog

* fix versions

* keep spinner while suggesting state

* remove clear button

* AI Assistant: change main action buttons style and behavior (#34370)

* optimize style declaration

* always use inputValue

* AI Assistant: add suggestions actions (#34399)

* add suggestion action buttons

* add editRequest state to handle back and forth editing 'mode'

* add discard callback on AI control, pass on tryAgain handler

* add changelog entries

* adjust behavior upon observations

* AI Assistant: improve spacing and sizes for block and action buttons (#34444)

* address spacing and size issues on the assistant block

* modify loading icon wrapper and size

* avoid height jumps when buttons wrapper is introduced

* fix spacing on icon buttons

* add changelogger entry for js-package

* AI Assistant: fix block to bottom when content exceeds the viewport height (#34452)

* fix ai block to bottom when content exceeds the viewport height

* add outside wrapper

* move positioning directive to outside wrapper

* adjust bottom space for outside wrapper

* AI Assistant: remove try again toolbar button (#34464)

* make handleTryAgain optional so to remove when not present

* add changelog entry

* AI Assistant bar: fix positioning on mobile (#34476)

* wrap assistant bar to control positioning

* always make sure the assistant bar is placed AFTER the toolbar

* add changelog entry

* AI Assistant: add prompt to attributes (#34438)

* add userPrompt to the block's attributes

* set last value to default to initial prop

* add requestingState to attributes to keep consistency between reloads

* add changelog entry

* AI assistant: change guideline component (#34496)

* fix multiple line display on ai control footer message

* export AI control message as FooterMessage

* add changelog entry

* AI Assistant: fix bottom padding issue (#34519)

* fix bottom padding by using bottom directive for sticky positioning

* add changelog entry

* remove unnecessary directive

* AI Assistant: restrict stored state value (#34522)

* restrict stored states to 'init' and 'done' so we don't reload into half way scenarios

* add changelog entry

* fix misuse of ternary

* add cancel button to get rid of the assistant block (#34524)

* AI Client: fix empty prompts handling (#34547)

* use a default value to keep track of changes in the prompt, allow for empty prompts to show suggestion action buttons

* use null instead of a made up default value

* handle edit mode logic on one-click actions (#34584)

* AI Assistant: prepare new UI/UX features for merge (#34620)

* consolidate changelog entry for ai-client

* consolidate changelog entry for Jetpack plugin

* add test instructions for the new AI Assistant UI/UX

* AI Assistant: change prompt placeholder (#34671)

* remove typing effect from prompt placeholders

* use a fixed placeholder with suggestions on how to create a prompt

* use different placeholders depending on whether there is content or not

* move ternary out of jsx chunk

* take i18n out of ternary

* change placeholder text

* remove back to edit button (#34675)

* add showRemove attribute to AIControl (#34677)

* Cleanup quick wins after feedback (#34679)

* clean up changelog entries for ai-client

* clean up changelog entries for jetpack plugin

* AI assistant: fix positioning (#34700)

* remove some confusing 100% widths messing with border alignments

* fix colliding bottom directives on assistant bar

* Membership services: small refactor (#34635)

* Very small clean up

* Rename abstract class filename

* Create membership-abstract-token-class

* Fix path

* fix typing

* Update class-wpcom-online-subscription-service.php

* Update projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-asbtract-token-subscription-service.php

Co-authored-by: Miguel Lezama <miguel@automattic.com>

* Update projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-jetpack-token-subscription-service.php

Co-authored-by: Miguel Lezama <miguel@automattic.com>

* Rename file

* Update class-wpcom-online-subscription-service.php

* Rename references to Token_Subscription_Service

* Fix naming

---------

Co-authored-by: Miguel Lezama <miguel@automattic.com>

* Boost: Refactor module, index and more into React (#34662)

* Create skeleton for Module component in React

* Refactor UpgradeCTA to React

* Added a small variant for toggle control

* Use small variant of toggle control

* Fix onEnable, onMountEnable, and onDisable

* Do not pass yearlyPricing

* Implement onEnable and onDisable

* changelog

* Create index component

* Make module errors consistent with old design

* Fix ISA alignment

* Improve alignments and spacing

* Add resize unavailable notice

* Add fake recommendations meta

* Add missing pieces

* Cleanup

* Remove Index.svelte

* Move pricing to ds based config

* Remove Module.svelte

* Use React DS based moduleState management

* Restore tsconfig change

* Change interface for useModuleState callbacks

* Call useModuleState hook directly on the component

* Fix infinite loop on purchase success

* Fixup project version

* Fix selector in test

* Make speed score dependent on useModulesState for auto-refresh to work

* Fix isModuleEnabled check

* Update modules-state endpoint in tests

---------

Co-authored-by: Mark George <thingalon@gmail.com>

* Boost: Migrate Tips to react (#34697)

* Update Tips to react

* add changelog

* Backport crm 6.3.2 Changes (#34708)

* Fix missing class fatal related to PR #34635 (#34705)

* CRM: start 6.3.3-alpha release cycle (#34710)

* Fix My plan dashboard js errors due to nested anchor tags  (#34707)

* Moved href from Button props to ExternalLink props to avoid duplicate anchor tags

* Added changelog

* Admin Page: avoid react warning with VideoPress card (#34713)

This should avoid us getting the following warning:

```
Warning: Failed prop type: Invalid prop `overrideContent` of type `boolean` supplied to `DashItem`, expected a single ReactElement.
DashItem@wp-content/plugins/jetpack/_inc/build/admin.js?ver=28b3a12169ab227bdcfa:36878:1
SettingsForm@wp-content/plugins/jetpack/_inc/build/admin.js?ver=28b3a12169ab227bdcfa:43524:3
ConnectFunction@wp-content/plugins/jetpack/_inc/build/admin.js?ver=28b3a12169ab227bdcfa:8870:68
ConnectFunction@wp-content/plugins/jetpack/_inc/build/admin.js?ver=28b3a12169ab227bdcfa:8870:68
DashVideoPress@wp-content/plugins/jetpack/_inc/build/admin.js?ver=28b3a12169ab227bdcfa:34467:1
```

* Fix missing class fatal related to PR #34635 (#34718)

* Subscriptions: update "verify your email" wall copy (#34716)

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

* Fix sharing buttons window open (#34703)

* Contact Form: add extra field settings to base field (#34704)

* Contact Form: add extra field settings to base field

* Fix package version

* Fix the media validation notice shown even when auto conversion is enabled. (#34730)

Co-authored-by: Gergely Juhasz <gergely.juhasz@automattic.com>

* Add `wpcom_ai_site_prompt` option to the site settings endpoint (#34709)

* Social: Fix endpoint backwards compatibility (#34566)

* Readd REST controllers with updated logic

* Readd class-settings classes for initial state population

* changelog

* Use depreacted and next_version tags

* Fixup versions

* Boost: Update Settings Page to react (#34714)

* Update Settings Page to react

* add changelog

* Sharing: update Twitter and X buttons to support official X button (#34719)

* Sharing: update Twitter and X buttons to support official X button

See #34459

- The official Twitter button is no more; it has been updated to the X design. Let's update our preview Twitter button match.
- Since we now have an official X button, let's offer it with the X service.

* Adjust official button styles in wp-admin

* This method is not needed

* Update domain name

* Update source name

* Bump version by specifying Jetpack version when enqueuing

* Update logo

* Update icon with more radius

* Social | Hide conversion notice for Simple sites (#34733)

* Hide notice on simple sites

* changelog

* Make auto-conversion disabled if not available

* Fixup versions

* Sharing buttons block alternate settings screen (#34673)

* build: Use a GitHub webhook to trigger the betadownload site (#34604)

Once Automattic/jetpack-builder#36 is deployed, we can update our build
script to produce the artifact for forked PRs and set up a GitHub
webhook for the `workflow_run` event to support testing with PRs from
forked repositories.

This requires the GitHub "Fork pull request workflows from outside
collaborators" setting to be set to "Require approval for all outside
contributors" to avoid unreviewed code potentially producing malicious
zips for the betadownload server. We've had that set since March 2023
(p9dueE-6Rp-p2).

This also updates the Jetpack Live Branches script to remove the check
for a forked PR, since we can now test with them.

* Contact Form: minify stylesheets in prod (#34672)

* Contact Form: minify stylesheets in prod

* Update registered styles paths

* Fix package version

* Tools: Add the Readme to the release script (#34722)

* Make sure the readme.txt file gets updated after saving changelog edits

* Fix command being run

* comment out a check

* pass the correct flag

* Undo comments

* Update the next steps at the end

* Fix token subscription service name (#34723)

* Rename Token_Subscript_Service to Abstract_Token_Subscription_Service

* changelog

* Allow old class just in case

* `use` classes at the top to shorten up the class check.

* Content Lens: move feature to `beta` (#34740)

* move `content-lens` feature to beta

* changelog

* Newsletter prepublish panel: update toggle appearance (#34724)

* Contact Form: add accessible name to form (#34667)

* Contact Form: add accessible name to form

* Update accessible name on content change

* Remove unused code

* Fix package version

* Add warning notice

* Backport mu-wpcom-plugin 2.0.8, jetpack 13.0-a.7 Changes (#34742)

* Changelog and readme.txt edits.

* Init new cycle

* Fix the stable tag

* [Boost] Fix importmap deferral (#34746)

* Add importmap to the set of script types we exclude from deferral

* changelog

* Fix versions in packages/forms

* changelog

---------

Co-authored-by: Mark George <thingalon@gmail.com>

* Update changelog (#34747)

* Admin page: remove alt attribute from decorative icon (#34715)

The Alt value isn't necessary here, since the icon doesn't bring any additional info on the page for folks using screen readers.

This has the added benefit of getting rid of a React warning:

```
Warning: Failed prop type: Invalid prop `alt` of type `object` supplied to `PlanIcon`, expected `string`.
PlanIcon@wp-content/plugins/jetpack/_inc/build/admin.js?ver=15a830895a6f6d1e3c39:44686:1
MyPlanCard@wp-content/plugins/jetpack/_inc/build/admin.js?ver=15a830895a6f6d1e3c39:51709:20
```

* Add/fix typo (#34712)

* Fix typo

* Chnagelog

---------

Co-authored-by: Fabien MILLERAND <fab.millerand@automattic.com>

* Boost: Refactor navigation into react (#34731)

* Use react router

* Add redirects

* Fix typo

* Add dummy ISA page

* Add tracks

* Restore getting started functionality

* Move props to DS

* Move connection to DS and cleanup props

* Remove connection from JS constants

* Improve getting started handling

* Remove wrapping individual components with `DataSyncProvider`

* Only start ISA if ISA is active

* changelog

* Use hook and component provided by react-router-dom

* Remove data-sync provider wrapping from settings-page

* Remove jetpack connection prop from analytics

Analytics isn't collected while there is no connection. So, this prop is unnecessary anyway.

* Pop bubbles for performanceHistoryNeedsUpdate

* Fix priority support

* Cast wpcom blogId into integer

* Fix getting started with free

* Fix getting started for real

* Show index for missing routes

* Stats: add passing select UTM parameters (#34431)

This commit adds select UTM parameters to the pixel requests send to wp.com, so that we can show more statistics to site owners.

---------

Co-authored-by: Gergely Márk Juhász <36671565+gmjuhasz@users.noreply.github.com>
Co-authored-by: valterlorran <valterlorran@hotmail.com>
Co-authored-by: Valter Lorran <valter.lorran@automattic.com>
Co-authored-by: Andrii Lysenko <60262784+andrii-lysenko@users.noreply.github.com>
Co-authored-by: Steve D <33553323+sdixon194@users.noreply.github.com>
Co-authored-by: Kev <kevin.zoschke@automattic.com>
Co-authored-by: arthur791004 <arthur.chu@automattic.com>
Co-authored-by: Mikael Korpela <mikael@ihminen.org>
Co-authored-by: Juanma Rodriguez Escriche <juanma.rodriguez.escriche@automattic.com>
Co-authored-by: Tim Broddin <tim@wannabes.be>
Co-authored-by: Samiff <samiff@users.noreply.github.com>
Co-authored-by: Karen Attfield <karenlattfield@gmail.com>
Co-authored-by: Calypso Bot <sysops+ghmatticbot@automattic.com>
Co-authored-by: Renovate Bot <bot@renovateapp.com>
Co-authored-by: Brad Jorsch <brad.jorsch@automattic.com>
Co-authored-by: Chris McCluskey <chris.mccluskey@automattic.com>
Co-authored-by: tbradsha <32492176+tbradsha@users.noreply.github.com>
Co-authored-by: Christian Gastrell <cgastrell@gmail.com>
Co-authored-by: MILLER/F <millerf@automattic.com>
Co-authored-by: Miguel Lezama <miguel@automattic.com>
Co-authored-by: Adnan Haque <3737780+haqadn@users.noreply.github.com>
Co-authored-by: Mark George <thingalon@gmail.com>
Co-authored-by: Peter Petrov <peter.petrov89@gmail.com>
Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>
Co-authored-by: Manzoor Wani <manzoorwani.jk@gmail.com>
Co-authored-by: Gergely Juhasz <gergely.juhasz@automattic.com>
Co-authored-by: Griffith Chen <cferrios@gmail.com>
Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>
Co-authored-by: Christopher Smith <8002138+chrisfromthelc@users.noreply.github.com>
Co-authored-by: Damián Suárez <rdsuarez@gmail.com>
Co-authored-by: thingalon <mark@thingalon.com>
Co-authored-by: Fabien MILLERAND <fab.millerand@automattic.com>
Co-authored-by: Piotr Stankowski <6437642+trakos@users.noreply.github.com>

Aurorum committed:
Revert "Catch up your branch (#2)" (#3)

This reverts commit 2434b7a4df9553f0627a49caa2043ec72fdca70b.

Aurorum committed:
Try force conflict

Aurorum committed:
Try force conflict again

Aurorum committed:
Merge branch 'trunk' into add/stats-block

Aurorum committed:
Remove forced conflict

Aurorum commented:
Welp, that didn't work, but thanks for giving it a go! Looks like I've made a mess of this branch in the process though - sorry about that. 

I've managed to update my branch again using GitHub, but I'm still getting the same error when committing my local changes - same with other branches (I tried with #34717 too). Any other ideas on where I might be going wrong? It was working fine earlier this week, and I don't believe that I've made any changes. :/ 

Thanks!

Aurorum committed:
Fix tests

Aurorum commented:
Sorry about the labels - I'm still having a bunch of trouble with my dev environment and trying to commit, so I can't run the tests locally. I think I've fixed them though, and that this branch is okay again.

anomiex commented:
The check compares `origin/trunk` with `HEAD`, and the trunk branch in your fork is 1226 commits behind Automattic/jetpack. Try syncing your fork; GitHub should give you a button for that when visiting https://github.com/Aurorum/jetpack/tree/trunk.

Aurorum committed:
Merge branch 'trunk' into add/stats-block

Aurorum commented:
Ah, that worked perfectly! Thank you.

Aurorum committed:
Fix tests

Aurorum committed:
Merge branch 'trunk' into add/stats-block

Aurorum commented:
Hi folks, just looping back on this old one and realised some tests were broken (sorry!). They should be fixed now, but I'd be grateful if someone could run them again so I can double-check. 

This should be good to review otherwise (sorry - I know the labels are a mess), but let me know if I'm missing anything else!

Aurorum committed:
Update composer.lock

Aurorum commented:
Looks like `composer.lock` still needed updating - should be taken care of now!

kangzj commented:
Thanks for working on this @Aurorum 👍 I'll take a look early tomorrow 😄

kangzj commented:
Sorry I didn't manage to get onto this one today. Will do first thing tomorrow.

Aurorum commented:
All good! :)

Aurorum committed:
Merge branch 'trunk' into add/stats-block

Aurorum committed:
Resolve conflicts

Aurorum committed:
Specific class names

kangzj commented on projects/packages/stats/src/class-wpcom-stats.php:
```diff
@@ -399,6 +404,44 @@ protected function fetch_stats( $args = array() ) {
+		if ( $stats_cache ) {
+			$data = reset( $stats_cache );
+
+			if ( is_wp_error( $data ) ) {
```
I took a look at the returning value of get_post_meta, which wouldn't be an array of WP_Error, so I guess the condition here will never be satisfied.

<strike>We could the block as a check whether $data is still an array; otherwise `key` function would throw out an error. So I guess we could do something like:

```
if( ! is_array($data) ) {
  return new WP_Error(...);
}
```
</strike>

Update:

I guess it might be better if let it continue without cache, if cache fails?

Aurorum replied:
> I took a look at the returning value of get_post_meta, which wouldn't be an array of WP_Error, so I guess the condition here will never be satisfied.

Ah, totally missed this. I've changed it now so that we only cache the stats data that we're interested in.

kangzj replied:
ah sorry! I would prefer the errors being cached please just in case. I was thinking whether WP_Error could be serialized and unserialized correctly, but I just tested, it worked fine!

Could we revert the change please?

Aurorum replied:
No problem - switched it back!

kangzj commented on projects/packages/stats/src/class-wpcom-stats.php:
```diff
@@ -203,13 +203,18 @@ public function get_file_downloads( $args = array() ) {
 		$this->resource = sprintf( 'post/%d', $post_id );
 
+		if ( $cache_in_meta ) {
+			$this->fetch_post_stats( $args, $post_id );
```
```suggestion
			return $this->fetch_post_stats( $args, $post_id );
```

I believe we should return the result here 😄

Aurorum replied:
Whoops, great spot! :)

kangzj reviewed (changes requested):
Thanks for your amazing work @Aurorum 👍 

All looks and works amazingly well except for missing a `return` and another nitpick 👍 👍 👍

kangzj committed:
Update projects/packages/stats/src/class-wpcom-stats.php

kangzj commented on projects/packages/stats/src/class-wpcom-stats.php:
```diff
@@ -399,6 +404,44 @@ protected function fetch_stats( $args = array() ) {
+		if ( $stats_cache ) {
+			$data = reset( $stats_cache );
+
+			if ( is_wp_error( $data ) ) {
+				return $data;
+			}
+
+			$time       = key( $data );
+			$expiration = self::STATS_CACHE_EXPIRATION_IN_MINUTES * MINUTE_IN_SECONDS;
+
+			if ( ( time() - $time ) < $expiration ) {
+				return array_merge( array( 'cached_at' => $time ), $data[ $time ] );
+			}
+		}
```
Maybe we could try something like:

```PHP
		if ( is_array( $data ) ) {
			  $time       = key( $data );
			  $expiration = self::STATS_CACHE_EXPIRATION_IN_MINUTES * MINUTE_IN_SECONDS;
  
			  if ( ( time() - $time ) < $expiration ) {
				  return array_merge( array( 'cached_at' => $time ), $data[ $time ] );
			  }
			}
		}
```

Aurorum replied:
> I guess it might be better if let it continue without cache, if cache fails?

If we're doing this, would it be preferable to just not cache the error instead? Something like this: 

```php
		$wpcom_stats = $this->fetch_remote_stats( $endpoint, $args );
		if ( ! is_wp_error( $wpcom_stats ) ) {
			update_post_meta( $post_id, $meta_name, array( time() => $wpcom_stats ) );
		}
```

Let me know if I'm misunderstanding you though! :) 

(And thank you for the review!!)

Aurorum replied:
I tested this a bit more and it seems to work well, so I've committed it now. That way we only cache the view count, which is the data we want to cache.

Happy to change it if you prefer a different approach though!

Aurorum committed:
Avoid caching error messages

Aurorum commented:
Thanks so much - very much appreciated! :))

Aurorum committed:
Revert caching change

kangzj reviewed (approved):
Looks good to me! Thanks for working on this 👍 ❤️

Aurorum commented:
Thanks so much for your help @kangzj! ❤️

Do we need to create an issue like #32937 on the steps still needed to get the block out of beta (supporting WPcom etc) or will that be okay?

kangzj commented:
> Do we need to create an issue like https://github.com/Automattic/jetpack/issues/32937 on the steps still needed to get the block out of beta (supporting WPcom etc)

Hi @Aurorum, thanks for your contribution. Yes, it'll be great to track it than forgetting it later 😄 I've filed an issue for that:

https://github.com/Automattic/jetpack/issues/35436

jeherve commented:
Related diff: D138696-code

Aurorum commented:
Thanks!! Out of interest, does that handle the WPcom side? If so, can I remove the WPcom check? 

https://github.com/Automattic/jetpack/blob/ce56219691155c4ab0941a31d1d6fa6d4f9367ec/projects/plugins/jetpack/extensions/blocks/blog-stats/blog-stats.php#L25

jeherve commented:
> does that handle the WPcom side? If so, can I remove the WPcom check?

We should indeed be able to remove that check now.

kangzj commented:
Thanks @jeherve 👍
